### PR TITLE
feat: add input validation utility functions

### LINKED ISSUE #915: Add input validation utility functions
## Description
Create reusable validation utility functions to standardize form validation across the application.

## Problem
Validation logic is duplicated across forms. We need centralized, reusable validation functions.

## Files to Create
- `src/utils/validators.ts`
- `src/utils/__tests__/validators.test.ts`

## Implementation

```typescript
// src/utils/validators.ts

/**
 * Validation result
 */
export interface ValidationResult {
  isValid: boolean;
  error?: string;
}

/**
 * Email validation
 */
export function validateEmail(email: string): ValidationResult {
  if (!email) {
    return { isValid: false, error: 'Email is required' };
  }
  
  const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
  
  if (!emailRegex.test(email)) {
    return { isValid: false, error: 'Invalid email format' };
  }
  
  return { isValid: true };
}

/**
 * Password validation
 */
export function validatePassword(password: string): ValidationResult {
  if (!password) {
    return { isValid: false, error: 'Password is required' };
  }
  
  if (password.length < 8) {
    return { isValid: false, error: 'Password must be at least 8 characters' };
  }
  
  if (!/[A-Z]/.test(password)) {
    return { isValid: false, error: 'Password must contain an uppercase letter' };
  }
  
  if (!/[a-z]/.test(password)) {
    return { isValid: false, error: 'Password must contain a lowercase letter' };
  }
  
  if (!/[0-9]/.test(password)) {
    return { isValid: false, error: 'Password must contain a number' };
  }
  
  return { isValid: true };
}

/**
 * Phone number validation (US format)
 */
export function validatePhone(phone: string): ValidationResult {
  if (!phone) {
    return { isValid: false, error: 'Phone number is required' };
  }
  
  const phoneRegex = /^\+?1?\s*\(?\d{3}\)?[\s.-]?\d{3}[\s.-]?\d{4}$/;
  
  if (!phoneRegex.test(phone)) {
    return { isValid: false, error: 'Invalid phone number format' };
  }
  
  return { isValid: true };
}

/**
 * URL validation
 */
export function validateUrl(url: string): ValidationResult {
  if (!url) {
    return { isValid: false, error: 'URL is required' };
  }
  
  try {
    new URL(url);
    return { isValid: true };
  } catch {
    return { isValid: false, error: 'Invalid URL format' };
  }
}

/**
 * Required field validation
 */
export function validateRequired(value: string, fieldName: string): ValidationResult {
  if (!value || value.trim() === '') {
    return { isValid: false, error: `${fieldName} is required` };
  }
  
  return { isValid: true };
}

/**
 * Min length validation
 */
export function validateMinLength(
  value: string,
  minLength: number,
  fieldName: string
): ValidationResult {
  if (value.length < minLength) {
    return {
      isValid: false,
      error: `${fieldName} must be at least ${minLength} characters`,
    };
  }
  
  return { isValid: true };
}

/**
 * Max length validation
 */
export function validateMaxLength(
  value: string,
  maxLength: number,
  fieldName: string
): ValidationResult {
  if (value.length > maxLength) {
    return {
      isValid: false,
      error: `${fieldName} must be no more than ${maxLength} characters`,
    };
  }
  
  return { isValid: true };
}

/**
 * Compose multiple validators
 */
export function composeValidators(
  ...validators: Array<() => ValidationResult>
): ValidationResult {
  for (const validator of validators) {
    const result = validator();
    if (!result.isValid) {
      return result;
    }
  }
  
  return { isValid: true };
}
```

## Usage Examples

```typescript
// In a form component
const handleSubmit = (e: FormEvent) => {
  e.preventDefault();
  
  const emailResult = validateEmail(email);
  if (!emailResult.isValid) {
    setEmailError(emailResult.error);
    return;
  }
  
  const passwordResult = validatePassword(password);
  if (!passwordResult.isValid) {
    setPasswordError(passwordResult.error);
    return;
  }
  
  // Proceed with submission
};

// Composing validators
const result = composeValidators(
  () => validateRequired(name, 'Name'),
  () => validateMinLength(name, 2, 'Name'),
  () => validateMaxLength(name, 50, 'Name')
);
```

## Tests
Create `src/utils/__tests__/validators.test.ts`:
```typescript
import {
  validateEmail,
  validatePassword,
  validatePhone,
  validateUrl,
  validateRequired,
  validateMinLength,
  validateMaxLength,
} from '../validators';

describe('validators', () => {
  describe('validateEmail', () => {
    it('should validate correct email', () => {
      expect(validateEmail('test@example.com').isValid).toBe(true);
    });
    
    it('should reject invalid email', () => {
      expect(validateEmail('invalid').isValid).toBe(false);
      expect(validateEmail('').isValid).toBe(false);
    });
  });
  
  describe('validatePassword', () => {
    it('should validate strong password', () => {
      expect(validatePassword('Password123').isValid).toBe(true);
    });
    
    it('should reject weak password', () => {
      expect(validatePassword('short').isValid).toBe(false);
      expect(validatePassword('nouppercase123').isValid).toBe(false);
    });
  });
  
  // ... more tests
});
```

## Tasks
- [ ] Create validators.ts utility file
- [ ] Implement email validation
- [ ] Implement password validation
- [ ] Implement phone validation
- [ ] Implement URL validation
- [ ] Implement required field validation
- [ ] Implement length validations
- [ ] Add compose validators helper
- [ ] Create comprehensive tests
- [ ] Add JSDoc comments
- [ ] Update 2-3 forms to use validators
- [ ] Document validation patterns

## Learning Outcomes
- Understanding validation patterns
- Learning about regex
- Practice with utility functions
- Understanding composition patterns

## Estimated Time
2-3 hours

## Priority
Medium - Improves code reusability

## Difficulty
Beginner-Intermediate

### PR DESCRIPTION
**Pull request by Mayeed-git**
## Description
Implements reusable validation utility functions to standardise form validation across the application.

### ✅ Core Utilities
- **Created `src/utils/validators.ts`** with 8 validation functions:
  - `validateEmail()` - Email format validation with regex
  - `validatePassword()` - Strong password requirements (8+ chars, uppercase, lowercase, number)
  - `validatePhone()` - US phone format validation (supports multiple formats)
  - `validateUrl()` - URL format validation using native URL constructor
  - `validateRequired()` - Non-empty field validation
  - `validateMinLength()` - Minimum length validation
  - `validateMaxLength()` - Maximum length validation
  - `composeValidators()` - Utility to chain multiple validators together

### ✅ Testing
- **Created `src/utils/__tests__/validators.test.ts`** with comprehensive test coverage:
  - 27 test cases covering all validators
  - Edge case testing (empty values, exact lengths, boundary conditions)
  - 100% code coverage
  - All tests passing ✅

### ✅ Integration
Updated 3 forms to use the new validators:
1. **NewsletterForm** - Replaced manual email regex with `validateEmail()`
2. **ContactForm** - Implemented `validateEmail()`, `validatePhone()`, and `validateRequired()` across 5 fields
3. **ApplyForm** - Integrated `validateEmail()` and `validateUrl()` for prework validation

### ✅ Documentation
- Added comprehensive JSDoc comments for all functions
- Included usage examples and parameter descriptions
- Clear return type documentation

## Checklist
- [x] Code follows project style guidelines
- [x] Added tests for all functions
- [x] Added JSDoc comments
- [x] No breaking changes

Closes #915

### CODE REVIEW

File: src/utils/validators.ts [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+ * validateEmail('test@example.com') // { isValid: true }
+ * validateEmail('invalid') // { isValid: false, error: 'Invalid email format' }
+ */
+
+export function validateEmail(email: string): ValidationResult {
+    if (!email) {
+        return { isValid: false, error: "Email is required" };
+    }
+
+    const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
```
[the remark this answers is not in the document]
Comment (jeromehardaway) [reply src/utils/validators.ts]: @Mayeed-git see above

File: src/utils/validators.ts [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+ * @example
+ * validatePhone('(555) 555-5555') // { isValid: true }
+ * validatePhone('123') // { isValid: false, error: 'Invalid phone number format' }
+ */
+export function validatePhone(phone: string): ValidationResult {
+    if (!phone) {
+        return { isValid: false, error: "Phone number is required" };
+    }
+
+    const phoneRegex = /^\+?1?\s*\(?\d{3}\)?[\s.-]?\d{3}[\s.-]?\d{4}$/;
```
[the remark this answers is not in the document]
Comment (jeromehardaway) [reply src/utils/validators.ts]: @Mayeed-git see above

File: src/components/forms/contact-form.tsx [hunk 3]
```diff
@@ -89,7 +93,10 @@ const ContactForm = forwardRef<HTMLFormElement, TProps>(({ className }, ref) =>
                         state={hasKey(errors, "phone") ? "error" : "success"}
                         showState={!!hasKey(errors, "phone")}
                         {...register("phone", {
-                            required: "Phone is required",
+                            validate: (value) => {
+                                const result = validatePhone(value);
+                                return result.isValid || result.error || "";
```
[the remark this answers is not in the document]
Comment (jeromehardaway) [reply src/components/forms/contact-form.tsx]: @Mayeed-git see above
